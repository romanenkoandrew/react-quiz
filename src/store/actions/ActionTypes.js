export const ActionTypes = {
  FETCH_QUIZES_START: "FETCH_QUIZES_START",
  FETCH_QUIZES_ERROR: "FETCH_QUIZES_ERROR",
  FETCH_QUIZES_SUCCESS: "FETCH_QUIZES_SUCCESS",
  FETCH_QUIZES_BY_ID_SUCCESS: "FETCH_QUIZES_BY_ID_SUCCESS",
  QUIZ_SET_STATE: "QUIZ_SET_STATE",
  FINISH_QUIZ: "FINISH_QUIZ",
  QUIZ_NEXT_QUESTION: "QUIZ_NEXT_QUESTION",
  RETRY_QUIZ: "RETRY_QUIZ",
  CREATE_QUIZ_QUESTION: "CREATE_QUIZ_QUESTION",
  RESET_QUIZ_CREATION: "RESET_QUIZ_CREATION",
  AUTH_SUCCESS: "AUTH_SUCCESS",
  AUTH_LOGOUT: "AUTH_LOGOUT",
};
