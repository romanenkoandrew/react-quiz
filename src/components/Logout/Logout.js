import React from "react";
import { connect } from "react-redux";
import { Redirect } from "react-router-dom";
import { logout } from "../../store/actions/auth";

export class Logout extends React.Component {
  componentWillUnmount() {
    this.props.logout();
  }
  render() {
    return <Redirect to="/" />;
  }
}

export const mapDispatchToProps = (dispatch) => {
  return {
    logout: () => dispatch(logout()),
  };
};
export default connect(null, mapDispatchToProps)(Logout);
