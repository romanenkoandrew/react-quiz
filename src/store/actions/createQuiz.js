import { ActionTypes } from "./ActionTypes";
import axios from "axios";
import { BASE_URL } from "../../Constants/Constants";

export function createQuizQuestion(item) {
  return {
    type: ActionTypes.CREATE_QUIZ_QUESTION,
    item,
  };
}
export function resetQuizCreation() {
  return {
    type: ActionTypes.RESET_QUIZ_CREATION,
  };
}
export function finishCreateQuiz() {
  return async (dispatch, getState) => {
    const state = getState().createQuiz;
    await axios.post(`${BASE_URL}quizes.json`, state.quiz);
    dispatch(resetQuizCreation());
  };
}
