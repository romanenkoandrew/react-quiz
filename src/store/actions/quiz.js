import axios from "axios";
import { BASE_URL } from "../../Constants/Constants";
import { ActionTypes } from "./ActionTypes";

export function fetchQuizes() {
  return async (dispatch) => {
    dispatch(fetchQuizesStart());
    try {
      const response = await axios.get(`${BASE_URL}quizes.json`);
      const quizes = [];
      Object.keys(response.data).forEach((key, index) => {
        quizes.push({ id: key, name: `Тест №${index + 1}` });
      });
      dispatch(fetchQuizesSuccess(quizes));
    } catch (e) {
      dispatch(fetchQuizesError(e));
    }
  };
}

export function fetchQuizById(id) {
  return async (dispatch) => {
    dispatch(fetchQuizesStart());
    try {
      const response = await axios.get(`${BASE_URL}quizes/${id}.json`);
      const quiz = response.data;
      dispatch(fetchQuizByIdSuccess(quiz));
    } catch (e) {
      dispatch(fetchQuizesError(e));
    }
  };
}

export function fetchQuizesStart() {
  return {
    type: ActionTypes.FETCH_QUIZES_START,
  };
}
export function fetchQuizesSuccess(quizes) {
  return {
    type: ActionTypes.FETCH_QUIZES_SUCCESS,
    quizes,
  };
}
export function fetchQuizByIdSuccess(quiz) {
  return {
    type: ActionTypes.FETCH_QUIZES_BY_ID_SUCCESS,
    quiz,
  };
}
export function fetchQuizesError(error) {
  return {
    type: ActionTypes.FETCH_QUIZES_ERROR,
    error,
  };
}

export function quizSetState(answerState, results) {
  return {
    type: ActionTypes.QUIZ_SET_STATE,
    answerState,
    results,
  };
}
export function finishQuiz() {
  return {
    type: ActionTypes.FINISH_QUIZ,
  };
}
export function quizNextQuestion(number) {
  return {
    type: ActionTypes.QUIZ_NEXT_QUESTION,
    number,
  };
}

export function retryQuiz() {
  return {
    type: ActionTypes.RETRY_QUIZ,
  };
}

export function quizAnswerClick(answerId) {
  return (dispatch, getState) => {
    const state = getState().quiz;
    if (state.answerState) {
      const key = Object.keys(state.answerState)[0];
      if (state.answerState[key] === "success") {
        return;
      }
    }
    const question = state.quiz[state.activeQuestion];
    const results = state.results;
    if (question.rightAnswerId === answerId) {
      if (!results[question.id]) {
        results[question.id] = "success";
      }
      dispatch(quizSetState({ [answerId]: "success" }, results));
      const timeout = window.setTimeout(() => {
        if (isQuizFinished(state)) {
          dispatch(finishQuiz());
        } else {
          dispatch(quizNextQuestion(state.activeQuestion + 1));
        }
        window.clearTimeout(timeout);
      }, 500);
    } else {
      results[question.id] = "error";
      dispatch(quizSetState({ [answerId]: "error" }, results));
    }
  };
  function isQuizFinished(state) {
    return state.activeQuestion + 1 === state.quiz.length;
  }
}
