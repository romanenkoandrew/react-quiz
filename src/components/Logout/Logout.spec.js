import React from "react";
import { shallow } from "enzyme";
import { Logout, mapDispatchToProps } from "./Logout";
import configureStore from "redux-mock-store";
import { ActionTypes } from "../../store/actions/ActionTypes";

const mockStore = configureStore();
const store = mockStore({});

describe("Logout component", () => {
  it("should render Logout without props", () => {
    const component = shallow(<Logout />);
    expect(component).toMatchSnapshot();
  });
  it("should logout", () => {
    const dispatch = jest.fn();

    mapDispatchToProps(dispatch).logout();
    expect(dispatch.mock.calls[0][0]).toEqual({
      type: ActionTypes.AUTH_LOGOUT,
    });
  });
  it("should call the mock logout function", () => {
    const mockLogoutFn = jest.fn();
    const component = shallow(<Logout logout={mockLogoutFn} />);
    component.unmount();
    expect(mockLogoutFn.mock.calls.length).toBe(1);
  });
});
