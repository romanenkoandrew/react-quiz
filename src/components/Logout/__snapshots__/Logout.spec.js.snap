// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logout component should render Logout without props 1`] = `
<Redirect
  to="/"
/>
`;
